[TheEnd-alr-taken] iP

TheEnd-alr-taken opened the pull request:
# Motiva: Task Management Chatbot  

> *"Your mind is for having ideas, not holding them."* – David Allen  

Motiva is a **personal task management chatbot** designed to help you keep track of your to-dos, deadlines, and events. It provides a **simple yet powerful** way to manage your daily tasks, featuring a **graphical user interface (GUI)** built with JavaFX.

## Features ✨

- [x] **Text-based interaction**
- [x] **Intuitive GUI with JavaFX**
- [x] **Task management (To-dos, Deadlines, Events)**

## How to Get Started 🚀

1. **Download the latest release** from [here](https://github.com/TheEnd-alr-taken/ip/releases).
2. **Run the application** using:  
   ```sh
   java -jar motiva.jar
   ```
3. **Add tasks** using simple commands.
4. **Enjoy a more organized life!** 😉

## Supported Commands 🔥

You can interact with Motiva using the following commands:

```
list                 		# Lists all tasks
bye                		# Exits the application
mark <index>        		# Marks a task as done
unmark <index>      		# Unmarks a task
delete <index>     		# Deletes a task
find <keyword>      		# Finds tasks containing the keyword
todo <task description>   			  # Adds a new to-do
deadline <task description> /by <date>  	  # Adds a new deadline
event <task description> /from <date> /to <date>  # Adds a new event
```

## Example Usage 🎯

```
> todo Read a book
[T] [ ] Read a book added!

> deadline Submit report /by 2024-03-15
[D] [ ] Submit report (by: Mar 15 2024) added!

> event Project meeting /from 2024-02-20 /to 2024-02-20
[E] [ ] Project meeting (from: Feb 20 2024 to: Feb 20 2024) added!

> list
1. [T] [ ] Read a book
2. [D] [ ] Submit report (by: Mar 15 2024)
3. [E] [ ] Project meeting (from: Feb 20 2024 to: Feb 20 2024)
```

## Why Use Motiva? 🤔

- **Simple & intuitive** 🧠
- **Built with JavaFX for a great UI** 🎨
- **Saves time and helps productivity** ⏳


## Development & Contribution 🛠️

To give a few comments as to how I can further improve this project of mine.


## About the Code 💻

The main entry point of the application is the `Main` class:

```java
public class Main extends Application {
    private Motiva motiva = new Motiva();

    @Override
    public void start(Stage stage) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/MainWindow.fxml"));
            AnchorPane ap = fxmlLoader.load();
            Scene scene = new Scene(ap);
            stage.setScene(scene);
            stage.setTitle("Motiva ChatBot");
            fxmlLoader.<MainWindow>getController().setMotiva(motiva);
            stage.show();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

## Future Plans 🔮

> _“Great things are done by a series of small things brought together.”_ – Vincent Van Gogh

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Commit:
Complete Level-0: Rename, Greet, Exit

Commit:
Complete Level-1: Echo

Commit:
Complete Level-2: Add, List

Commit:
Complete Level-3: Mark as Done

Commit:
Complete Level-4: ToDo, Event, Deadline

Commit:
Complete A-TextUiTesting: Automated Text UI Testing

Commit:
Complete Level-5: Handle Errors

Commit:
Complete Level-6: Delete

Commit:
Complete Level-7: Save

Commit:
Merge branch 'branch-Level-7'

Commit:
Complete Level-8: Dates and Times

Commit:
Merge branch 'branch-Level-8'

Commit:
Complete A-MoreOOP: Use More OOP

Mil-leon commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,23 @@
+        super(description);
+        this.by = super.parseDateTime(by);
+    }
+
```
Perhaps some comment's to elaborate on what the functions do would be nice 😄

Mil-leon commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,188 @@
+                                + "\ttodo <task description>\n"
+                                + "\tdeadline <task description> /by <due date>\n"
+                                + "\tevent <task description> /from <fromDate> /to <toDate>\n";
+                Ui.formatReply("Invalid command. Please try one of the following commands:\n" + commands);
```
Really like how all the possible commands are listed out in the case of an invalid command

Mil-leon commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,78 @@
+    protected String description;
+    protected boolean isDone;
+
+    protected static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    protected static final DateTimeFormatter DATE_TIME_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    protected static final DateTimeFormatter DISPLAY_FORMAT = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
```
Like how the data stored for each Task is static and final

Mil-leon commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,31 @@
+
+        for (String line : lines) {
+            System.out.println(" "+ indent + line);
+        }
+
+        System.out.println(indent + "============================================================\n");
```
Perhaps can make a method to print this seperator exclusively?

Mil-leon reviewed (commented):
Overall good work. Code was short and concise and I really like how you utilise regex to parse the data. Could probably create more classes to seperate the commands. :)

Commit:
Complete A-Packages: Organize into Packages

Commit:
Merge branch 'add-gradle-support'

Commit:
Complete A-Gradle: Use Gradle

Commit:
Complete A-JUnit: Add JUnit Tests

Commit:
Complete Level-9: Find

Commit:
Complete A-CodingStandard: Following the Coding Standard

Commit:
Merge branch 'branch-Level-9'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'master' into branch-A-JavaDoc

Commit:
Complete A-JavaDoc: JavaDoc

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Complete Level-10: GUI

Commit:
Merge branch 'branch-Level-10'

Commit:
Complete A-CheckStyle: Use CheckStyle

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add assertions and GUI improvements, fix checkstyle issues

- Added assertions in key methods:
  - parseDateTime: Ensure valid date format is used.
  - handleUserInput: Assert motiva is not null.
  - formatReply: Ensure Motiva's reply is never empty.

- Made improvements to the GUI.

- Fixed checkstyle issues to align with Java coding standards.

These changes ensures critical conditions are met, improving the
user interface, and maintaining code quality.

Commit:
Refactor Storage and Parser for improved code quality

Split loadFromStorage() method in Storage.java into smaller, more
manageable methods for better readability and maintainability.

Create parseTaskFromLine() method in Parser.java to cater for
parsing of task from a given file line.

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Change content of data file

No other major changes to the code quality. This commit is done
to rectify previous branching mistake.

TheEnd-alr-taken committed:
Merge pull request #2 from TheEnd-alr-taken/branch-A-CodeQuality

Branch a code quality

Commit:
Implement sorting feature for tasks

Add sorting functionality to TaskList:
  - Deadlines are sorted chronologically by their `by` field.
  - Events are sorted chronologically by their `from` field.
  - Todos are placed at the end.

Update parseCommand() in Parser.java to support the sort command.

This feature improves task organization, making it easier for users
to view deadlines and events in a chronologically order.

Commit:
Merge branch 'branch-C-Sort'

Commit:
Fix error display when loading tasks from file fails

Commit:
Complete A-UserGuide: User Guide
